### ISSUE TITLE
expval instructions in the circuit

### ISSUE THREAD

**Issue by ajavadia** [Qiskit/qiskit#11643]
### What should we add?

This is for discussion.

In the Estimator, it is necessary to track circuits and observables (and parameters) together. "PUB" (primitive unit block) has been proposed for this. However the transpiler does not handle PUB. It does handle circuits and parameters, and so allowing it to modify observables basically means it will be able to modify PUBs.

Here are a few use cases:
- the transpiler often transpiles circuits up to some final permutation. Recently we added a `observable.apply_layout()` method. But this requires the user to actively call this otherwise they get a wrong output. The user has to basically always remember to write `observable.apply_layout(transpiled_circuit.layout)` after every `transpile()`.
- the transpiler might want to do some clifford manipulations in the circuit. This has a straightforward correspondence with Pauli observables. If we do not allow the transpiler to accordingly update the observable, again the user has to do it (there is no corresponding `circuit.layout` which says the circuit has been transpiled up to some final clifford, so we might need to supply this).
- perhaps my circuit has some ZZ(theta) gates in the last layer, and I'm measuring Z observables. Then those gates can be dropped without affecting the observable. Again the compiler cannot know this without knowing what is to be measured.
- as a user I want to use Aer's Estimator to simulate expectation values of a circuit. However Aer is not smart enough to know that a 100-qubit circuit where only 10 qubits are active and only 10-qubit Paulis are being measured can be simulated. It raises an error. It's Sampler does not raise this error I believe, because it just looks at which qubits are measured. But when it looks at the observable it sees a 100-qubit Pauli. Of course this can be mitigated by making Aer smarter about handling `I` Paulis. 

With all of the above, I think we might want to consider having `expval` instructions in the circuit. This would make the observable more explicit and make the circuit "complete", in the sense that what output we want computed from it. 

The alternative is to do what we have been doing so far which is to have ad-hoc code in different places in case we need to handle circuit+observable together. This includes users also handling the pair together.

**Comment by valleyofblackpanther** [comment]
Hi, @ajavadia I have a hypothetical approach to implement something like this, i don't if its right, correct me if I am wrong. 
In Qiskit, we typically measure an observable by first transforming the circuit such that the observable is diagonal in the computational basis, and then performing measurements. The `expval` instruction would be a higher-level abstraction that automates this process. 
This is some sort of pseudocode for an `expval` instruction to a `QuantumCircuit` within Qiskit. I am assuming that there is a addition of a new `Expval` class and the necessary modifications to the `QuantumCircuit` class to include the `expval` method:

```
from qiskit.circuit import QuantumCircuit, Instruction
from qiskit.circuit.library.standard_gates import IGate
from qiskit.quantum_info import Pauli, Operator
from qiskit.dagcircuit import DAGCircuit  

class Expval(Instruction):
    """Expectation value instrcution"""
    def __init__(self, observable, qubits):
        #initialize the instruction with the observable and the target qubits
        super().__init__("expval", len(qubits), 0, [observable])
        self.observable = observable 
        self.qubits = qubits

def apply_expval(self, observable, qubits):
    """Apply the expectation value instruction to the circuit"""
    if isinstance(observable, Pauli):
        #convert the Pauli observable to a circuit
        circ = observable.to_circuit()
    elif isinstance(observable, Operator):
        #Decompose the operator into a basis of Pauli operators
        #this is a placeholder for the actual decomposition logic
        raise NotImplementedError("Operator decomposition is not yet implemented.") 
    else:
        raise ValueError("Unsupported observable type: {}".format(type(observable)))
    
    #create the expval instruction
    expval_instruction = Expval(observable, qubits)

    #append the instruction to the circuit
    self.append(expval_instruction, qubits)

#extend the QuantumCircuit class with the expval method
QuantumCircuit.expval = apply_expval   
```
The code is not complete or runnable because it lacks backend execution logic and the detailed mechanics of transforming the circuit according to the observable.
But according to me, to implement something like this, it requires the following steps to be taken:
1. Defining the new instruction
2. Circuit integration
3. Transpiler integration
4. Backend execution
5. Result Processing
6. Unit testing

**Comment by valleyofblackpanther** [comment]
@ajavadia Can i be assigned to this issue?

**Comment by jakelishman** [comment]
@valleyofblackpanther: thanks for your interest, but this issue is not ready for implementation yet.  The questions raised in the original comment are not so much about _how_ this should be done, but rather whether the design fits within the data model of `QuantumCircuit`, how it interacts with the responsibilities of the primitives, and what if any classical values it would represent within the circuit.

The next steps for this issue are likely to be a more fleshed out design proposal by @ajavadia in Qiskit/RFCs, and technical review on whether this is a direction we want to move in.

**Comment by valleyofblackpanther** [comment]
Okay @jakelishman thanks for letting me know 👍🏾

**Comment by jaygambetta** [comment]
I agree with bringing observables closer to circuits as so many things transpire the circuit also affect the observables. I also believe as we extend observables to effects (non-hermitian) then this gives us a lot of power to the transpilier to do both together.
